### PR TITLE
Collect per module eBPF helper telemetry

### PR DESCRIPTION
**Pull request by usamasaqib**
### What does this PR do?
This PR collects eBPF helper error telemetry for each module in its own maps.

The PR also removes the dependency on the ebpf-manager for the `ebpfTelemetry` type. This will allow modules not using the ebpf-manager to manage the lifecycle of their eBPF programs, to use this framework.

The `ebpfTelemetry` struct now tracks telemetry maps for each module, sized according to the number of maps and probes the module's manager is attempting to load. The collector loops over all the maps for all the tracked modules to emit the telemetry.

### Motivation
1. Size the eBPF telemetry maps correctly  based on the number of maps and programs being tracked.
2. Breakdown error telemetry by module.
3. Support dynamicinstrumentation use case where we cannot predict any suitable default value for map sizes.

### Describe how to test/QA your changes
Changes are already covered by functional tests run with KMT.

### CODE REVIEW

File: pkg/ebpf/telemetry/modifier.go [hunk 1]
```diff
@@ -22,16 +38,105 @@ func (t *ErrorsTelemetryModifier) String() string {
...
+	}
+
+	mapErrMap, _, err := m.GetMap(mapErrTelemetryMapName)
+	if err != nil {
+		return fmt.Errorf("failed to find map %s: %w", mapErrTelemetryMapName, err)
+	}
+	genericMapErrMap, err := maps.Map[uint64, mapErrTelemetry](mapErrMap)
+	if err != nil {
+		return fmt.Errorf("failed to create generic map from map spec for %s: %w", mapErrTelemetryMapName, err)
+	}
```
Comment (gjulianm) [inline pkg/ebpf/telemetry/modifier.go:126]: There's a shorthand in the `maps` package for this:

```suggestion
	genericMapErrMap, err := maps.GetMap[uint64, mapErrTelemetry](m, mapErrTelemetryName)
	if err != nil {
		return fmt.Errorf("failed to get map %s: %w", mapErrTelemetryMapName, err)
	}
```

File: pkg/ebpf/telemetry/modifier.go [hunk 2]
```diff
@@ -22,16 +38,105 @@ func (t *ErrorsTelemetryModifier) String() string {
...
+
+	helperErrMap, _, err := m.GetMap(helperErrTelemetryMapName)
+	if err != nil {
+		return fmt.Errorf("failed to find map %s: %w", helperErrTelemetryMapName, err)
+	}
+	genericHelperErrMap, err := maps.Map[uint64, helperErrTelemetry](helperErrMap)
+	if err != nil {
+		return fmt.Errorf("failed to create generic map from map spec for %s: %w", helperErrTelemetryMapName, err)
 	}
```
Comment (gjulianm) [inline pkg/ebpf/telemetry/modifier.go:135]: Same as above

**Review by gjulianm — APPROVED** [review]
LGTM, just some minor comments

**Review by guyarb — COMMENTED** [review]
Is there any impact of the memory of the ebpf maps, instruction count for the different programs, or anything else noticable?

**Comment by usamasaqib** [comment]
> Is there any impact of the memory of the ebpf maps, instruction count for the different programs, or anything else noticable?

This approach increases the memory used by 32KB for `map_err_telemetry_map`.
image

However, it reduces the memory consumption for `helper_err_telemetry_map` by 102KB.
image

Going forward the approach will cause slightly more memory usage due to the extra metadata associated with the new eBPF maps, however, overall the memory usage maybe less due to more accurate sizing of the maps.

There is no effect on instruction count.

**Review by guyarb — APPROVED** [review]
LGTM for USM

**Review by hmahmood — APPROVED** [review]

File: pkg/ebpf/manager.go [hunk 3]
```diff
@@ -66,23 +66,26 @@ type Modifier interface {
...
-	BeforeInit(*manager.Manager, names.ModuleName, *manager.Options, io.ReaderAt) error
+	BeforeInit(*manager.Manager, names.ModuleName, *manager.Options) error
 
 	// AfterInit is called after the ebpf.Manager.InitWithOptions call
 	AfterInit(*manager.Manager, names.ModuleName, *manager.Options) error
 }
 
 // InitWithOptions is a wrapper around ebpf-manager.Manager.InitWithOptions
 func (m *Manager) InitWithOptions(bytecode io.ReaderAt, opts *manager.Options) error {
+	if err := m.LoadELF(bytecode); err != nil {
```
Comment (val06) [inline pkg/ebpf/manager.go:77]: can you add a comment to state why this is necessary **before** calling to all the modifiers` `BeforeInit`?

**Review by val06 — APPROVED** [review]
left some comments

File: pkg/ebpf/manager.go [hunk 4]
```diff
@@ -14,13 +14,15 @@ import (
 
 	manager "github.com/DataDog/ebpf-manager"
 
+	"github.com/DataDog/datadog-agent/pkg/ebpf/names"
 	"github.com/DataDog/datadog-agent/pkg/util/log"
 )
 
 // Manager wraps ebpf-manager.Manager, adding a property with the list of enabled modifiers
```
[the remark this answers is not in the document]
Comment (usamasaqib) [reply pkg/ebpf/manager.go:21]: This is documented in the package itself.

**Pull request merged**